swarm_dial.go

```diff
@@ -529,3 +593,48 @@ func (s *Swarm) dialAddr(ctx context.Context, p peer.ID, addr ma.Multiaddr) (tra
 	// success! we got one!
 	return connC, nil
 }
+
+// TODO We should have a `IsFdConsuming() bool` method on the `Transport` interface in go-libp2p-core/transport.
+// This function checks if any of the transport protocols in the address requires a file descriptor.
+func (s *Swarm) IsFdConsumingAddr(addr ma.Multiaddr) bool {
+	noProxyAddrFnc := func(addr ma.Multiaddr) bool {
+		tpt := s.TransportForDialing(addr)
+		if tpt == nil {
+			// let's return true to be safe
+			log.Debugf("no transport registered for address %s, so marking it as Fd consuming for safety", addr.String())
+			return true
+		}
+		protos := tpt.Protocols()
+		// is a fd consuming transport present in the list of protocols that the transport supports ?
+		for i := range protos {
+			if _, ok := fdConsumingTptProtos[protos[i]]; ok {
+				return true
+			}
+		}
+		return false
+	}
```

Comment:
Hm. I take back what I said before. Let's just simplify this:

1. If the transport is a "proxy" transport, don't consume a file descriptor token.
2. Otherwise, just call `addrutil.IsFDCostlyTransport`.

Reply:
@Stebalien Please see above comment.

For dialing in the limiter, we do NOT consume fd's for proxy addrs.
For classifying an address as FD or Non-FD consuming so we can dial all the latter ones before attempting the former, we look at the address of the relay server.

Reply:
Ah, I see. You're right, we do need to do that.

Reply:
In that case, I'd still simplify this (given that we only really support circuit addresses anyways):

1. Modify `addrutil.IsFDCostlyTransport` to ignore circuit addresses.
2. When sorting between relay connections when dialing, split on the `P_CIRCUIT` protocol and call `addrutil.IsFDCostlyTransport` on the relay part.

That way:

* We don't need to look at the swarm's transports. You're right in that we should probably expose this information on the transport itself, but we don't do that yet anyways.
* Can just go "full special case" for circuit addresses.

Does it sound like that will simplify this?

Reply:
@aarshkshah1992 thoughts on this?

Reply:
@Stebalien I'd replied earlier at https://github.com/libp2p/go-libp2p-swarm/pull/212#issuecomment-626501013.

Reproducing it here:

"We shouldn't leak how our dial uses the addrutil.IsFDCostlyTransport utility into the 
shared utility itself. I've documented the code better for now and have filed an issue(#214 ) to simplify/fix this later.
".

Reply:
I still think this code needs to be simpler:

* We're checking to see if the address is a "proxy" address, but we only _actually_ support circuit addresses. We should just split on the circuit address and be done with it.
* We're hard coding rules for specific transport protocols when we should be able to just say "if protocol X appears in the target multiaddr (before any proxy protocols), the address must consume a file descriptor". That is, we know that the tcp and unix protocols both _must_ consume a file descriptor per connection while udp, memory, etc. protocols don't.

Basically, this code is trying to be general purpose and abstract over transports, but we can't actually do that so it's hard coding a bunch of stuff. Sitting halfway in between is just confusing.